File: 2022/Python/day_3.py
def parse_input():
    splitter = []
    with open("Inputs/day_3_input.txt") as input_file:
        for line in input_file:
            ltext = line.replace("\n", "")
            splitter.append(ltext)
    return splitter


def day_1():
    data = parse_input()
    sumr = 0
    alpha_lower = "abcdefghijklmnopqrstuvwxyz"
    alpha_upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    for i in data:
        s1 = list(i[: int(len(i) / 2)])
        s2 = list(i[int(len(i) / 2) :])

        item_match = ""

        for j in s1:
            if j in s2:
                item_match = j
                break

        i_index = alpha_lower.find(item_match)
        if i_index == -1:
            i_index = alpha_upper.find(item_match) + 27
        else:
            i_index += 1

        sumr += i_index

    print("The sum of priorties for Part 1 is: " + str(sumr))


def day_2():
    data = parse_input()
    sumr = 0
    alpha_lower = "abcdefghijklmnopqrstuvwxyz"
    alpha_upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    # Same as above but read 3 lines at once and treat each as their own (don't split the string)
    for i in range(0, len(data), 3):
        letter = ""
        l_1 = list(data[i])
        l_2 = list(data[i + 1])
        l_3 = list(data[i + 2])

        # Find longest
        if len(l_1) >= len(l_2) and len(l_1) >= len(l_3):
            longest = l_1
        elif len(l_2) > len(l_1) and len(l_2) >= len(l_3):
            longest = l_2
        else:
            longest = l_3

        for j in longest:
            if j in l_1 and j in l_2 and j in l_3:
                letter = j
                break

        i_index = alpha_lower.find(letter)
        if i_index == -1:
            i_index = alpha_upper.find(letter) + 27
        else:
            i_index += 1

        sumr += i_index

    print("The sum of priorties for Part 2 is: " + str(sumr))


if __name__ == "__main__":
    day_1()
    day_2()

File: 2022/cpp/day_3.cpp
#include <fstream>
#include <iostream>
#include <vector>

void part_1();
void part_2();
std::vector<std::string> parse_input();

void part_1() {
  std::vector<std::string> day_data = parse_input();
  int sumr = 0;
  std::string alpha_lower = "abcdefghijklmnopqrstuvwxyz";
  std::string alpha_upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

  for (size_t i = 0; i < day_data.size(); i++) {
    std::string x = day_data[i];
    int x_size = x.length() / 2;
    std::string s1 = x.substr(0, x_size);
    std::string s2 = x.substr(x_size, x_size);

    char item_match;
    for (size_t j = 0; j < s1.length(); j++) {
      char find_s = s1[j];
      int find_loc = s2.find(find_s);

      if (find_loc > -1) {
        item_match = find_s;
        break;
      }
    }

    int find_l = alpha_lower.find(item_match);
    if (find_l < 0) {
      find_l = alpha_upper.find(item_match) + 27;
    } else {
      find_l += 1;
    }

    sumr += find_l;
  }

  std::cout << "The answer to Part 1 is : " << sumr << std::endl;
}
void part_2() {
  std::vector<std::string> day_data = parse_input();
  int sumr = 0;
  std::string alpha_lower = "abcdefghijklmnopqrstuvwxyz";
  std::string alpha_upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

  for (size_t i = 0; i < day_data.size(); i += 3) {
    std::string s1 = day_data[i];
    std::string s2 = day_data[i + 1];
    std::string s3 = day_data[i + 2];
    std::string longest;
    char item_match;

    if (s1.length() > s2.length() && s1.length() > s3.length()) {
      longest = s1;
    } else if (s2.length() > s1.length() && s2.length() > s3.length()) {
      longest = s2;
    } else {
      longest = s3;
    }

    for (size_t j = 0; j < longest.length(); j++) {
      char checker = longest[j];
      int find_1 = s1.find(checker);
      int find_2 = s2.find(checker);
      int find_3 = s3.find(checker);

      if (find_1 > -1 && find_2 > -1 && find_3 > -1) {
        item_match = checker;
        break;
      }
    }

    int find_l = alpha_lower.find(item_match);
    if (find_l < 0) {
      find_l = alpha_upper.find(item_match) + 27;
    } else {
      find_l += 1;
    }

    sumr += find_l;
  }

  std::cout << "The answer to Part 2 is : " << sumr << std::endl;
}

int main() {
  part_1();
  part_2();
  return 0;
}

std::vector<std::string> parse_input() {
  std::vector<std::string> nums;
  std::ifstream f("Inputs/day_3_input.txt");

  std::string s;

  while (getline(f, s)) {
    nums.push_back(s);
  }

  f.close();

  return nums;
}
